pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,363 @@
+		"endpoint", "discover",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+	ws, err := s.upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to upgrade request to websocket connection")
+		return err
+	}
+	defer ws.Close()
+
+	_, data, err := ws.ReadMessage()
+	if err != nil {
+		logger.WithError(err).Warn("Failed to read message")
+		return err
+	}
+	var req messages.DiscoverQuery
+	if err := json.Unmarshal(data, &req); err != nil {
+		logger.WithError(err).Warn("Failed to parse discover query message")
+		return err
+	}
+
+	if req.EUI.IsZero() {
+		writeDiscoverError(s.ctx, ws, "Empty router EUI provided")
+		return errEmptyGatewayEUI
+	}
+
+	ids := ttnpb.GatewayIdentifiers{
+		EUI: &req.EUI.EUI64,
+	}
+	ctx, ids, err = s.server.FillGatewayContext(ctx, ids)
+	if err != nil {
+		writeDiscoverError(ctx, ws, fmt.Sprintf("Failed to fetch gateway: %s", err.Error()))
+		return err
+	}
+
+	scheme := "ws"
+	if c.IsTLS() {
+		scheme = "wss"
+	}
+	res := messages.DiscoverResponse{
+		EUI: req.EUI,
+		Muxs: basicstation.EUI{
+			Prefix: "muxs",
+		},
+		URI: fmt.Sprintf("%s://%s%s", scheme, c.Request().Host, c.Echo().URI(s.handleTraffic, ids.GatewayID)),
+	}
+	data, err = json.Marshal(res)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to marshal response message")
+		writeDiscoverError(ctx, ws, "Router not provisioned")
+		return err
+	}
+	if err := ws.WriteMessage(websocket.TextMessage, data); err != nil {
+		logger.WithError(err).Warn("Failed to write discover response message")
+		return err
+	}
+	logger.Debug("Sent discover response message")
+	return nil
+}
+
+func (s *srv) handleTraffic(c echo.Context) error {
+	var latestUpstreamXTime int64
+	id := c.Param("id")
+	var auth string
+	bearerAuth := c.Request().Header.Get("Bearer")
+	if bearerAuth != "" {
+		auth = bearerAuth
+	} else {
+		auth = c.Request().Header.Get(echo.HeaderAuthorization)
+	}
+	ctx := c.Request().Context()
+	if auth != "" {
+		md := metadata.New(map[string]string{
+			"id":            id,
+			"authorization": fmt.Sprintf("Bearer %s", auth),
+		})
+		if ctxMd, ok := metadata.FromIncomingContext(s.ctx); ok {
+			md = metadata.Join(ctxMd, md)
+		}
+		ctx = metadata.NewIncomingContext(s.ctx, md)
+	}
+
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "traffic",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+
+	ctx, ids, err := s.server.FillGatewayContext(ctx, ttnpb.GatewayIdentifiers{GatewayID: id})
+	if err != nil {
+		return err
+	}
+
+	// For gateways with valid EUIs and no auth, we provide the link rights ourselves as in the udp frontend.
+	if auth == "" {
+		ctx = rights.NewContext(ctx, rights.Rights{
+			GatewayRights: map[string]*ttnpb.Rights{
+				id: {
+					Rights: []ttnpb.Right{ttnpb.RIGHT_GATEWAY_LINK},
+				},
+			},
+		})
+	}
+
+	uid := unique.ID(ctx, ids)
+	ctx = log.NewContextWithField(ctx, "gateway_uid", uid)
+
+	fp, err := s.server.GetFrequencyPlan(ctx, ids)
```

Comment:
This call is intended for gateways themselves requesting the frequency plan, like gRPC frontend. It's quite an expensive call goes it hits IS (including auth check) and the frequency plan store (by default GitHub).

This call is also performed by `Connect()` so let's do it once. To make FP available here, please expose it through a getter call on `Connection`